### PR TITLE
Add ConsoleLink CRD to enable custom links in help and user dropdowns

### PR DESCRIPTION
**Pull request by rhamilto**
Screen Shot 2019-04-23 at 11 46 24 AM
Screen Shot 2019-04-23 at 11 47 32 AM
Screen Shot 2019-04-23 at 11 47 25 AM

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: consolelinks.console.openshift.io
  annotations:
    displayName: ConsoleLinks
    description: Extension for customizing OpenShift web console links
spec:
  scope: Cluster
  group: console.openshift.io
  versions:
  - name: v1
    served: true
    storage: true
  names:
    plural: consolelinks
    singular: consolelink
    kind: ConsoleLink
    listKind: ConsoleLinkList
  additionalPrinterColumns:
  - name: Text
    type: string
    JSONPath: .spec.text
  - name: URL
    type: string
    JSONPath: .spec.href
  - name: Menu
    type: string
    JSONPath: .spec.menu
  - name: Age
    type: date
    JSONPath: .metadata.creationTimestamp
  subresources:
    status: {}
  validation:
    openAPIV3Schema:
      properties:
        spec:
          type: object
          description: Represents console link customizations spec
          required:
          - text
          - href
          - location
          properties:
            text:
              type: string
              description: Text of the link
            href:
              type: string
              description: Absolute secure URL for the link (must use https)
              pattern: '^https://([\w-]+.)+[\w-]+(/[\w- ./?%&=])?$'
            location:
              type: string
              description: Determines which dropdown menu the link is added to (HelpMenu, UserMenu)
              pattern: ^(HelpMenu|UserMenu)$
```

### CODE REVIEW

File: frontend/public/actions/features.ts [hunk 1]
```diff
@@ -156,6 +156,18 @@ const detectUser = dispatch => coFetchJSON('api/kubernetes/apis/user.openshift.i
     },
   );
 
+const detectConsoleLinks = dispatch => coFetchJSON('api/kubernetes/apis/console.openshift.io/v1/consolelinks')
+  .then(
```
Comment (benjaminapetersen) [inline frontend/public/actions/features.ts]: @spadgett @rhamilto curious if this has been tested as a non-clusteradmin/kubeadmin user.  I don't think normal users will be able to see these.

File: frontend/public/components/masthead-toolbar.jsx [hunk 2]
```diff
@@ -174,6 +174,10 @@ class MastheadToolbar_ extends React.Component {
     window.open(window.SERVER_FLAGS.requestTokenURL, '_blank').opener = null;
   }
 
+  _getAdditionalLinks(links, type) {
```
Comment (benjaminapetersen) [inline frontend/public/components/masthead-toolbar.jsx]: I'll doc in the API resource that they will be ordered alphabetically.

Comment (spadgett) [reply frontend/public/components/masthead-toolbar.jsx]: Do we need to clarify? If we doc that, we might not be able to change it later

File: frontend/public/actions/features.ts (hunk 1 above)
Comment (spadgett) [reply frontend/public/actions/features.ts]: We will need to add a role binding for authenticated users to list/get/watch console links

**Comment by benjaminapetersen** [comment]
The CRD file can be removed from this PR. `frontend/extensions/console-link.crd.yaml`

File: frontend/public/actions/features.ts (hunk 1 above)
Comment (benjaminapetersen) [reply frontend/public/actions/features.ts]: RBAC added here: https://github.com/openshift/console-operator/pull/229/files#diff-917b39b291003c5a9d8a57afceaa65aaR68
With an objection, but no surprise: https://github.com/openshift/console-operator/pull/229/files#r288233402

File: frontend/public/components/masthead-toolbar.jsx (hunk 2 above)
Comment (benjaminapetersen) [reply frontend/public/components/masthead-toolbar.jsx]: Works, I'll not then.

**Comment by christianvogt** [comment]
@spadgett we should create extensions for all these areas of contribution and then have the CRD simply populate the extensions
FYI @vojtechszocs

**Comment by spadgett** [comment]
> @spadgett we should create extensions for all these areas of contribution and then have the CRD simply populate the extensions
> FYI @vojtechszocs

Makes sense, I agree.

If you don't mind, I suggest we get the initial change in and then make it a proper extension in a follow-on so we don't miss 4.2. This PR has been blocked for about 3 months on API changes and was originally written before static plugins were added!

File: frontend/public/models/yaml-templates.ts [hunk 3]
```diff
@@ -808,6 +808,15 @@ spec:
...
     name: worker
+`).setIn([referenceForModel(k8sModels.ConsoleLinkModel), 'default'], `
+apiVersion: console.openshift.io/v1
+kind: ConsoleLink
+metadata:
+  name: example
+spec:
+  href: 'https://www.example.com'
+  location: HelpMenu
+  text: Additional help menu link
```
Comment (spadgett) [inline frontend/public/models/yaml-templates.ts]: nit: We probably want title case in the YAML template since the other menu items are title case

**Comment by spadgett** [comment]
> @spadgett we should create extensions for all these areas of contribution and then have the CRD simply populate the extensions
> FYI @vojtechszocs

@christianvogt I opened an issue to make this change: https://jira.coreos.com/browse/CONSOLE-1571

**Pull request merged**